### ISSUE TITLE
Mac: wrong behavior when closing the app window

### ISSUE THREAD

**Issue by acbalingit** [pbek/QOwnNotes#1098]
Typical app behavior on Macs is that pressing the close button will close the window but put the app in the background, and the there is a separate menu item (or pressing cmd+Q) to close the app entirely. In QOwnNotes, pressing the close button closes the entire app. I'm assuming the app follows the close button behavior in Windows.

While `Show menu bar item` partially solves this issue by keeping the app running in the menu bar, opening the window by clicking the dock icon doesn't work. On some Mac apps I encountered, they offer an option to hide the dock icon while closing the app when the menu bar item is still up. 

I'm not sure if this is an issue in Mac implementation or a limitation to Qt; might be good to just put this as enhancement. 

#### Expected behaviour
Pressing the close button on the window closes the only the window without quitting the app. When `Show menu bar item` is enabled, the window should be brought back up by clicking the dock icon. 

#### Actual behaviour
Pressing the close button quits the app and disappears from the dock. When enabling `Show menu bar item`, the button closes the app to background and keeps the icon on the dock but clicking the icon doesn't open the app again; you need to open it from the menu bar to bring it back up.

#### Steps to reproduce
On a Mac, press the close button
#### Output from the debug section in the settings dialog
**Current Date**: `Thu Dec 27 19:09:33 2018`
**Version**: `18.12.6`
**Build date**: `Dec 26 2018`
**Build number**: `4005`
**Platform**: `macosx`
**Operating System**: `macOS 10.14`
**Build architecture**: `x86_64`
**Current architecture**: `x86_64`
**Release**: `Travis CI`
**Qt Version (build)**: `5.12.0`
**Qt Version (runtime)**: `5.12.0`
**Portable mode**: `no`
...
**Qt Debug**: `0`
...
**Locale (interface)**: *empty*
**Icon theme**: `breeze-dark-qownnotes`
**Notes in current note folder**: `12`
**Enabled scripts**: `0`
...
#### Relevant log output in the Log panel
```
=== STARTING THE APP ===
[Dec 27 19:07:18] [debug]: QNetworkReplyHttpImplPrivate::_q_startOperation was called more than once QUrl("https://p.qownnotes.org/piwik.php?idsite=5&_id=308df8314da51302&cid=308df8314da51302&url=http://QOwnNotes/app/start&rec=1&apiv=1&res=1680x1050&ua=Mozilla/5.0 (Macintosh macOS 10.14; en-ph) PiwikTracker/0.1 (Qt/5.12.0 )&lang=en-ph&dimension1=18.12.6&dimension2=en_PH&dimension3=0&dimension7=macOS 10.14 (x86_64)&dimension9=Travis CI&action_name=App Start")
[Dec 27 19:07:18] [debug]: on_noteTreeWidget_currentItemChanged
[Dec 27 19:07:18] [debug]: setCurrentNote  - 'note':  Note: <id>1 <name>"Test Note" <fileName>"Test Note.md" <noteSubFolderId>0 <relativePath>"" <hasDirtyData>false  - 'updateNoteText':  true  - 'updateSelectedNote':  false
[Dec 27 19:07:18] [debug]:  added to history:  NoteHistoryItem: <noteName>"Test Note" <noteSubFolderPathData>"" <cursorPosition>0 <relativeScrollBarPosition>0
[Dec 27 19:07:18] [info]: "QOwnNotes server listening on port 22222"
[Dec 27 19:07:18] [debug]: Reply from  "/index.php/apps/qownnotesapi/api/v1/note/app_info"
[Dec 27 19:07:18] [debug]: Reply from app version test
[Dec 27 19:07:18] [debug]: Reply from  "/ocs/v1.php/apps/files_sharing/api/v1/shares"
[Dec 27 19:07:18] [debug]: Reply from share api
[Dec 27 19:07:18] [debug]: updateNoteShareStatusFromShare  - 'status':  "ok\n"
=== ON CLOSE BUTTON PRESS ===
[Dec 27 19:07:32] [debug]: storeSettings  - 'noteHistoryItem':  NoteHistoryItem: <noteName>"Test Note" <noteSubFolderPathData>"" <cursorPosition>4065 <relativeScrollBarPosition>0.942857
```

**Comment by pbek** [comment]
The only way I could mimic that would be to minimize the window in macOS if the close button is clicked. That seems to have the same effect (but shows a minimize animation).

I've found no other way so far.

Would that hurt the macOS feelings too much?

**Comment by pbek** [comment]
@Nitish, did you remove your comment? I can't see it anymore.

## 18.12.9
- on macOS the application will now be minimized when closed instead of quit to
  more mimic the default behavior of applications on macOS

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if it works for you?

**Comment by acbalingit** [comment]
@pbek release works on my end, but it does indeed just imitates the minimize behavior.

I tried looking for QT5 apps that does the "hide on close, open on dock icon click" behavior and the best I got is qBittorrent. In their implementation they set `app.setQuitOnLastWindowClosed(false)` and added a handler for dock icon click. See:  
* https://github.com/qbittorrent/qBittorrent/blob/master/src/gui/macutilities.mm#L52
* https://github.com/qbittorrent/qBittorrent/blob/master/src/gui/mainwindow.cpp#L1344
* https://github.com/qbittorrent/qBittorrent/blob/master/macxconf.pri#L10

where their implementation is mostly based on this StackOverflow answer:
* https://stackoverflow.com/questions/15143369/qt-on-os-x-how-to-detect-clicking-the-app-dock-icon

Sadly it adds a lot of stuff from the default behavior (which QOwnNotes does now) which makes maintainability a bit hard, but this is a really nice QoL improvement for macOS users' feelings. I don't have the time right now but I can make a pull request if it permits. The links above should be enough reference for implementing the behavior.

Also, thanks for looking into this. It's just a nitpick but again, it really will make the Mac app behavior consistent Mac apps and make it a *really good* Evernote replacement.
